hadoop-common-project/hadoop-auth/pom.xml

```diff
@@ -272,5 +272,68 @@
         </plugins>
       </build>
     </profile>
+    <profile>
+      <id>java-8</id>
+      <activation>
+        <jdk>1.8</jdk>
+      </activation>
+      <properties>
+        <maven.compiler.source>${javac.version}</maven.compiler.source>
+        <maven.compiler.target>${javac.version}</maven.compiler.target>
+      </properties>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-enforcer-plugin</artifactId>
+            <version>3.0.0</version>
+            <executions>
+              <execution>
+                <id>enforce-java-version</id>
+                <goals>
+                  <goal>enforce</goal>
+                </goals>
+                <configuration>
+                  <rules>
+                    <requireJavaVersion>
+                      <version>1.8</version>
+                    </requireJavaVersion>
+                  </rules>
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+    <profile>
+      <id>x509-suppress</id>
+      <activation>
+        <jdk>[9,)</jdk>
+      </activation>
+      <properties>
+        <maven.compiler.source>${javac.version}</maven.compiler.source>
+        <maven.compiler.target>${javac.version}</maven.compiler.target>
```

Comment:
Looks like these properties are set regardless of Java version.  So they could be moved out of the profiles.

However, we can use the profiles to set `maven.compiler.release` for Java 9+ instead of the other two properties.

Java 9 introduced new [`javac` option `--release`](https://docs.oracle.com/en/java/javase/17/docs/specs/man/javac.html#option-release), which helps ensure compatibility with the given Java version ([blog with great explanation](https://www.morling.dev/blog/bytebuffer-and-the-dreaded-nosuchmethoderror/)).  The goal of this change is to set this option when compiling with JDK 9+, instead of `--source` and `--target`.

`--release` expects version without the old-style `1.` prefix (i.e. `8` instead of `1.8`).  `maven-enforcer-plugin` [added support for the same](https://issues.apache.org/jira/browse/MENFORCER-440) in version 3.2.1, so we also need to bump the plugin.

Reply:
do you want to leverage the `--release` to allow Hadoop building against Java 17 but could run with Java 8 and above?

Reply:
`source` and `target` are set for the same reason, to be able to build with newer Java, but still create binaries for Java 8.  `release` is similar, but better.

Reply:
I'm not sure it really works. I have experienced a lot of the following error cases when run jar built with newer JDK with release 8

```
java.lang.NoSuchMethodError: java.nio.ByteBuffer.limit(I)Ljava/nio/ByteBuffer
```

Reply:
> I have experienced a lot of the following error cases when running jar on JDK 8 that built with newer JDK with release 8

Using `-target` or `--release`?

Because that's exactly the kind of problem `--release` solves compared to `-target`.

It's easy to reproduce the problem and verify the fix:

```bash
$ cat > NoSuchMethod.java <<EOF
import java.nio.ByteBuffer;

public class NoSuchMethod {

  public static void main(String[] args) {
    System.out.println(ByteBuffer.allocate(100).position(20).limit(62).remaining());
  }

}
EOF

$ usr/lib/jvm/java-11-openjdk-amd64/bin/javac -Xlint:none -source 8 -target 8 NoSuchMethod.java \
  && /usr/lib/jvm/java-8-openjdk-amd64/bin/java NoSuchMethod
Exception in thread "main" java.lang.NoSuchMethodError: java.nio.ByteBuffer.position(I)Ljava/nio/ByteBuffer;
	at NoSuchMethod.main(NoSuchMethod.java:6)

$ /usr/lib/jvm/java-11-openjdk-amd64/bin/javac --release 8 NoSuchMethod.java \
  && /usr/lib/jvm/java-8-openjdk-amd64/bin/java NoSuchMethod
42
```

Reply:
Wow, thanks for letting me know it can be resolved by `--release`

Reply:
that bytebuffer thing is trouble as even recent java8 open jdks have that bytebuffer override; you need to build on an oracle jdk right now. this will be an improvement if we can trust it